Add --replace flag to intercept command

petergardfjall opened the pull request:
## Description

This PR adds a `--replace` flag to the `telepresence intercept` command which, when specified, will replace application containers in the intercepted workload.

This PR builds on the server-side machinery implemented in https://github.com/telepresenceio/telepresence/pull/3330.

Having the possibility to replace workload containers should provide a solution for tickets like https://github.com/telepresenceio/telepresence/issues/1646 and https://github.com/telepresenceio/telepresence/issues/1608.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.yml`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

petergardfjall committed:
Add --replace flag to intercept command

Signed-off-by: Peter Gardfjäll <peter.gardfjall.work@gmail.com>

thallgren commented:
Hi @petergardfjall . Thanks for providing this. Can you please sign your commit and also please fix the [lint complaint](https://github.com/telepresenceio/telepresence/actions/runs/7346076447/job/20000253186?pr=3469)?

petergardfjall commented:
> Hi @petergardfjall . Thanks for providing this. Can you please sign your commit and also please fix the [lint complaint](https://github.com/telepresenceio/telepresence/actions/runs/7346076447/job/20000253186?pr=3469)?

Hi @thallgren, I've addressed your comments.

One thing I noted is that in its present shape `--replace` won't preserve volume mounts of the original workload pod at the local mount point. I suppose that is to be expected since the `replace` will simply delete the workload pod (it is implemented by the [deleteAppContainer](https://github.com/telepresenceio/telepresence/blob/3155355294f4dfe1fa3762b83decd13ee4491c65/cmd/traffic/cmd/manager/mutator/agent_injector.go#L199) patch if I'm reading it right). I suppose that patch would need to do finer-grained surgery on the application containers if the intercept is to preserve volume mounts. Perhaps something along the lines of:

- replace `image` with `k8s.gcr.io/pause` (or similar)
- remove `readinessProbe` and `livenessProbe`

That is perhaps a bit more similar in spirit to the `swap-deployment` command in Telepresence v1?

Either way I think that should not pollute this PR.
If you can provide some guidance I might take a stab at it in a follow-up PR.

thallgren commented:
Agree about not polluting this PR. If mounts doesn't work, then that's a separate issue. Not sure why that doesn't work. It really should, even when the container is replaced.

The `readinessProbe`/`livenessProbe` intended for the deleted container is also something that needs to be fixed. Not sure about replacing the image with `k8s.gcr.io/pause` though. The agent really should take over whatever it's doing (including mounts).

Commit:
Add integration test for the telepresence intercep --replace flag

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

thallgren commented:
This option is already present in our enterprise edition, so rather than asking you to provide an integration test, I pushed a commit to your branch with the test that we have already have.

petergardfjall commented:
> Agree about not polluting this PR. If mounts doesn't work, then that's a separate issue.

Opened an issue https://github.com/telepresenceio/telepresence/issues/3483 with a proposed fix in https://github.com/telepresenceio/telepresence/pull/3484.
